src/coreclr/src/gc/unix/cgroup.cpp

```diff
@@ -394,20 +465,120 @@ class CGroup
+        long long period;
+
+        quota = ReadCpuCGroupValue(CGROUP1_CFS_QUOTA_FILENAME);
+        if (quota <= 0)
+            return false;
+
+        period = ReadCpuCGroupValue(CGROUP1_CFS_PERIOD_FILENAME);
+        if (period <= 0)
+            return false;
+
+        ComputeCpuLimit(period, quota, val);
+
+        return true;
+    }
+
+    static bool GetCGroup2CpuLimit(uint32_t *val)
+    {
+        char *filename = nullptr;
+        FILE *file = nullptr;
+        char *endptr = nullptr;
+        char *max_quota_string = nullptr;
+        char *period_string = nullptr;
+        char *context = nullptr;
+        char *line = nullptr;
+        size_t lineLen = 0;
+
+        long long quota = 0;
+        long long period = 0;
+
+        bool result = false;
+
+        if (s_cpu_cgroup_path == nullptr)
+            return false;
+
+        if (asprintf(&filename, "%s%s", s_cpu_cgroup_path, CGROUP2_CPU_MAX_FILENAME) < 0)
+            return false;
+
+        file = fopen(filename, "r");
+        if (file == nullptr)
+            goto done;
+
+        if (getline(&line, &lineLen, file) == -1)
+            goto done;
+
+        // The expected format is:
+        //     $MAX $PERIOD
+        // Where "$MAX" may be the string literal "max"
+
+        max_quota_string = strtok_r(line, " ", &context);
+        if (max_quota_string == nullptr)
+        {
+            assert(!"Unable to parse " CGROUP2_CPU_MAX_FILENAME " file contents.");
+            goto done;
+        }
+        period_string = strtok_r(nullptr, " ", &context);
+        if (period_string == nullptr)
+        {
+            assert(!"Unable to parse " CGROUP2_CPU_MAX_FILENAME " file contents.");
+            goto done;
+        }
+
+        // "max" means no cpu limit
+        if (strcmp("max", max_quota_string) == 0)
+            goto done;
+
+        errno = 0;
+        quota = strtoll(max_quota_string, &endptr, BASE_TEN);
+        if (max_quota_string == endptr || errno != 0)
+            goto done;
+
+        period = strtoll(period_string, &endptr, BASE_TEN);
+        if (period_string == endptr || errno != 0)
+            goto done;
+
+        ComputeCpuLimit(period, quota, val);
+        result = true;
+
+    done:
+        if (file)
+            fclose(file);
+        free(filename);
+        free(line);
+
+        return result;
+    }
+
+    static void ComputeCpuLimit(long long period, long long quota, uint32_t *val)
+    {
+        // Cannot have less than 1 CPU
+        if (quota <= period)
+        {
+            *val = 1;
+            return;
+        }
+
+        // Calculate cpu count based on quota and round it up
+        double cpu_count = (double) quota / period  + 0.999999999;
+        *val = (cpu_count < UINT32_MAX) ? (uint32_t)cpu_count : UINT32_MAX;
+    }
+
     static long long ReadCpuCGroupValue(const char* subsystemFilename){
         char *filename = nullptr;
         bool result = false;
-        long long val;
+        long long val = 0;
```

Comment:
A nit - this is not necessary, the `val` is not used until it is successfully set by the `ReadLongLongValueFromFile`.